pkg/vgmanager/cleanup.go

```diff
@@ -0,0 +1,119 @@
+	if err := r.Client.List(ctx, lvs, &client.ListOptions{Limit: int64(1)}); err == nil {
+		// do not proceed with cleanup if there are logicalVolumes
+		if len(lvs.Items) > 0 {
+			return fmt.Errorf("found presence of logicalvolumes and aborting volume group cleanup")
+		}
+	} else {
+		r.Log.Error(err, "failed to list logicalvolumes CRs")
+		return err
+	}
+
+	// get all the volume groups from lvmd file
+	lvmdConfig := &lvmCMD.Config{
+		SocketName: controllers.DefaultLVMdSocket,
+	}
+
+	cfgBytes, err := os.ReadFile(controllers.LvmdConfigFile)
+	if os.IsNotExist(err) {
+		r.Log.Info("lvmd config file doesn't exist, no need of cleanup")
+		return nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to read lvmd config file", "file path", controllers.LvmdConfigFile)
+		return err
+	} else {
+		err = yaml.Unmarshal(cfgBytes, &lvmdConfig)
+		if err != nil {
+			r.Log.Error(err, "failed to unmarshal yaml contents of lvmd config data")
+			return err
+		}
+	}
+
+	// remove volume group sequentially
+	var done int
+	isStarted := false
+	for done = 0; done < len(lvmdConfig.DeviceClasses); done++ {
+		deviceClass := lvmdConfig.DeviceClasses[done]
+		if err = RemoveVolumeGroup(deviceClass.Name); err != nil {
+			if errors.Is(err, ErrNotFound) {
+				// TODO: simplify the logic, multiple controllers are mutating same
+				// LVMCluster CR and as a result during the reconciliation cleanup is
+				// happening but failing to remove finalizer. So, if we try to remove
+				// a VG and it isn't found then that was probably deleted in previous
+				// reconile and so mark that the cleanup process was initiated
+				isStarted = true
+			}
+			r.Log.Error(err, "failed to remove volume group", "deviceClass", deviceClass.Name)
+			break
+		}
+		if !isStarted {
+			isStarted = true
+		}
+	}
+
+	if !isStarted && len(lvmdConfig.DeviceClasses) > 0 {
+		// not able to start cleanup, no need to overwrite config file, just return error
+		return fmt.Errorf("failed to initiate cleanup of Volume Groups")
+	} else if done == len(lvmdConfig.DeviceClasses)-1 || len(lvmdConfig.DeviceClasses) == 0 {
+		// every volume group is deleted and config file can be deleted
+		if err = os.Remove(controllers.LvmdConfigFile); err != nil {
+			return fmt.Errorf("failed to remove LVMd config file")
+		}
+	} else if done != 0 {
+		// some vgs are pending for deletion, overwrite config file
+		lvmdConfig.DeviceClasses = lvmdConfig.DeviceClasses[done:]
+		config, err := yaml.Marshal(lvmdConfig)
+		if err != nil {
+			r.Log.Error(err, "failed to marshal yaml contents of lvmd config data")
+			return err
+		}
+		if err = os.WriteFile(controllers.LvmdConfigFile, config, 0600); err != nil {
+			return fmt.Errorf("failed to overwrite LVMd config file")
+		}
+		pending := make([]string, 0, len(lvmdConfig.DeviceClasses))
+		for _, dc := range lvmdConfig.DeviceClasses {
+			pending = append(pending, dc.Name)
+		}
+		r.Log.Info("device classes are pending for deletion", "deviceClasses", pending)
```

Comment:
Do we really need another list `pending` over here or can be just use `lvmdConfig.DeviceClasses` again?  
```suggestion
		r.Log.Info("device classes are pending for deletion", "deviceClasses", lvmdConfig.DeviceClasses)
```